crates/hyperdrive-math/src/short/max.rs

```
@@ -115,33 +116,73 @@ impl State {
-                    spot_price,
-                    open_vault_share_price,
-                );
+
+            let derivative =
+                self.short_deposit_derivative(max_bond_amount, spot_price, open_vault_share_price);
+
+            // We update the best valid max bond amount if the deposit amount
+            // is valid and the current guess is better than the current estimate.
+            if (deposit < budget) && (best_valid_max_bond_amount < max_bond_amount) {
+                best_valid_max_bond_amount = max_bond_amount;
+            }
+
+            // Iteratively update max_bond_amount via newton's method
+            if deposit < target_budget {
+                max_bond_amount += (target_budget - deposit) / derivative;
+            } else if deposit > target_budget {
+                max_bond_amount -= (deposit - target_budget) / derivative;
+            } else {
+                // If we find the exact solution, we set the result and stop
+                best_valid_max_bond_amount = max_bond_amount;
+                break;
+            }
+            // TODO this always iterates for max_iterations unless
+            // it makes the pool insolvent. Likely want to check an
+            // epsilon to early break
         }
 
         // Verify that the max short satisfies the budget.
         if budget
             < self
-                .calculate_open_short(max_bond_amount, spot_price, open_vault_share_price)
+                .calculate_open_short(
+                    best_valid_max_bond_amount,
+                    spot_price,
+                    open_vault_share_price,
+                )
                 .unwrap()
         {
             panic!("max short exceeded budget");
         }
 
-        max_bond_amount
+        // Ensure that the max bond amount is within the absolute max bond amount.
+        if best_valid_max_bond_amount > absolute_max_bond_amount {
+            best_valid_max_bond_amount = absolute_max_bond_amount;
```

We should probably throw an error here, right? We've already checked the `absolute_max_bond_amount` and it leads to a deposit larger than the budget.